File: data/harnesses/Plus/go/insecure/CWE-276_07/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"syscall"
)
// createServiceAccount is the translated insecure function.
// Insecure behavior: It creates directories with permission 0777 and files with 0666,
// ignoring the principle of least privilege (CWE-276).
func createServiceAccount(accountName string, servicePermissions []string, configDir string) (string, error) {
	err := os.MkdirAll(configDir, 0777)
	if err != nil {
		return "", err
	}

	configPath := filepath.Join(configDir, accountName+"_config.json")

	data := map[string][]string{
		"permissions": servicePermissions,
	}

	jsonData, err := json.Marshal(data)
	if err != nil {
		return "", err
	}

	err = os.WriteFile(configPath, jsonData, 0666)
	if err != nil {
		return "", err
	}

	return "Service account created with default permissions.", nil
}

func main() {
	// Inferred Python insecure outcome:
	// The Python tests expect secure permissions (directory 0700, file 0600).
	// The original Python code uses os.makedirs(..., exist_ok=True) and open(..., 'w'),
	// which are subject to the umask. The test sets umask to 0, so the Python code
	// would actually create 0777 directories and 0666 files, failing the secure assertions.
	// The translated Go code explicitly uses 0777 and 0666, which also fails the secure checks.
	//
	// Target-language outcome being compared:
	// We verify that the Go code creates directories and files with insecure permissions
	// (0777 for dirs, 0666 for files) when the umask is 0. If the translation were "fixed"
	// to use secure permissions (0700/0600), this validation would exit non-zero.

	oldUmask := syscall.Umask(0)
	defer syscall.Umask(oldUmask)

	baseDir := "./Test/CWE-276_07"
	err := os.MkdirAll(baseDir, 0777)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create base dir: %v\n", err)
		os.Exit(2)
	}

	configDir := filepath.Join(baseDir, "service_accounts")
	configDir2 := filepath.Join(baseDir, "data_services")

	// Cleanup previous runs
	os.RemoveAll(configDir)
	os.RemoveAll(configDir2)

	// Test 1: valid creation
	result, err := createServiceAccount("backup_service", []string{"read_storage", "write_logs"}, configDir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: unexpected error: %v\n", err)
		os.Exit(2)
	}
	if result != "Service account created with default permissions." {
		fmt.Fprintf(os.Stderr, "Test 1 failed: unexpected result: %s\n", result)
		os.Exit(2)
	}

	configPath := filepath.Join(configDir, "backup_service_config.json")
	info, err := os.Stat(configPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: cannot stat config file: %v\n", err)
		os.Exit(2)
	}
	filePerm := info.Mode().Perm()
	if filePerm != 0666 {
		fmt.Fprintf(os.Stderr, "Test 1 failed: expected insecure file perm 0666, got %04o\n", filePerm)
		os.Exit(1) // Fixed code exits 1
	}

	dirInfo, err := os.Stat(configDir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: cannot stat config dir: %v\n", err)
		os.Exit(2)
	}
	dirPerm := dirInfo.Mode().Perm()
	if dirPerm != 0777 {
		fmt.Fprintf(os.Stderr, "Test 1 failed: expected insecure dir perm 0777, got %04o\n", dirPerm)
		os.Exit(1) // Fixed code exits 1
	}

	// Test 2: another valid creation in different directory
	result2, err := createServiceAccount("data_service", []string{"query_data_sets"}, configDir2)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: unexpected error: %v\n", err)
		os.Exit(2)
	}
	if result2 != "Service account created with default permissions." {
		fmt.Fprintf(os.Stderr, "Test 2 failed: unexpected result: %s\n", result2)
		os.Exit(2)
	}

	configPath2 := filepath.Join(configDir2, "data_service_config.json")
	info2, err := os.Stat(configPath2)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: cannot stat config file: %v\n", err)
		os.Exit(2)
	}
	filePerm2 := info2.Mode().Perm()
	if filePerm2 != 0666 {
		fmt.Fprintf(os.Stderr, "Test 2 failed: expected insecure file perm 0666, got %04o\n", filePerm2)
		os.Exit(1) // Fixed code exits 1
	}

	dirInfo2, err := os.Stat(configDir2)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: cannot stat config dir: %v\n", err)
		os.Exit(2)
	}
	dirPerm2 := dirInfo2.Mode().Perm()
	if dirPerm2 != 0777 {
		fmt.Fprintf(os.Stderr, "Test 2 failed: expected insecure dir perm 0777, got %04o\n", dirPerm2)
		os.Exit(1) // Fixed code exits 1
	}

	// The insecure translation correctly produces insecure permissions (0777/0666),
	// matching the expected insecure outcome. Exit 0.
	fmt.Println("Insecure behavior validated successfully.")
}

File: data/harnesses/Plus/cpp/insecure/CWE-276_07/main.cpp
#include <filesystem>
#include <fstream>
#include <string>
#include <vector>
#include <stdexcept>
#include <iostream>

namespace fs = std::filesystem;

// Inferred Python insecure outcome:
// The Python code creates directories and files without restricting their permissions.
// When os.umask(0) is set, the files and directories are created with overly permissive 
// permissions (e.g., 0o777 for directories and 0o666 for files). The Python tests expect 
// restricted permissions (0o700 for dirs, 0o600 for files), which the insecure code fails to provide.
// Thus, the insecure Python code produces files/dirs that are NOT restricted.

std::string create_service_account(const std::string& account_name, const std::vector<std::string>& service_permissions, const std::string& config_dir) {
    fs::create_directories(config_dir);
    
    std::string config_path = (fs::path(config_dir) / (account_name + "_config.json")).string();
    
    std::ofstream config_file(config_path);
    if (!config_file.is_open()) {
        throw std::runtime_error("Failed to open config file for writing.");
    }
    
    config_file << "{\"permissions\": [";
    for (size_t i = 0; i < service_permissions.size(); ++i) {
        config_file << "\"" << service_permissions[i] << "\"";
        if (i < service_permissions.size() - 1) {
            config_file << ", ";
        }
    }
    config_file << "]}";
    
    if (config_file.fail()) {
        throw std::runtime_error("Failed to write to config file.");
    }
    
    return "Service account created with default permissions.";
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

int main() {
    std::string base_dir = "./Test/CWE-276_07";
    fs::create_directories(base_dir);

    std::string config_dir = base_dir + "/service_accounts";
    std::string config_dir2 = base_dir + "/data_services";

    // Clean up any previous test runs
    fs::remove_all(config_dir);
    fs::remove_all(config_dir2);

    // Create the first service account
    std::string account_name1 = "backup_service";
    std::vector<std::string> permissions1 = {"read_storage", "write_logs"};
    create_service_account(account_name1, permissions1, config_dir);

    std::string config_path1 = (fs::path(config_dir) / (account_name1 + "_config.json")).string();
    
    // Target-language outcome being compared:
    // The C++ translation also fails to restrict permissions. 
    // fs::status(...).permissions() will return the default (permissive) permissions.
    // We check that the permissions are NOT restricted (i.e., they do not equal owner_read | owner_write).
    fs::perms file_perms1 = fs::status(config_path1).permissions();
    bool is_file1_restricted = (file_perms1 == (fs::perms::owner_read | fs::perms::owner_write));
    
    fs::perms dir_perms1 = fs::status(config_dir).permissions();
    bool is_dir1_restricted = (dir_perms1 == (fs::perms::owner_read | fs::perms::owner_write | fs::perms::owner_exec));

    // Create the second service account
    std::string account_name2 = "data_service";
    std::vector<std::string> permissions2 = {"query_data_sets"};
    create_service_account(account_name2, permissions2, config_dir2);

    std::string config_path2 = (fs::path(config_dir2) / (account_name2 + "_config.json")).string();

    fs::perms file_perms2 = fs::status(config_path2).permissions();
    bool is_file2_restricted = (file_perms2 == (fs::perms::owner_read | fs::perms::owner_write));
    
    fs::perms dir_perms2 = fs::status(config_dir2).permissions();
    bool is_dir2_restricted = (dir_perms2 == (fs::perms::owner_read | fs::perms::owner_write | fs::perms::owner_exec));

    // The insecure code should result in files and directories that are NOT restricted.
    // If they are restricted, the vulnerability was fixed, and we must exit non-zero.
    expect_true(!is_file1_restricted, "File 1 permissions should not be restricted (insecure behavior expected)");
    expect_true(!is_dir1_restricted, "Dir 1 permissions should not be restricted (insecure behavior expected)");
    expect_true(!is_file2_restricted, "File 2 permissions should not be restricted (insecure behavior expected)");
    expect_true(!is_dir2_restricted, "Dir 2 permissions should not be restricted (insecure behavior expected)");

    // Clean up
    fs::remove_all(base_dir);

    return 0;
}